consensus/view_change_construct.go

```diff
@@ -0,0 +1,422 @@
+		}
+	}
+
+	// check when M3 sigs > M2 sigs, then M1 (recvMsg.Payload) should not be empty
+	preparedBlock := &types.Block{}
+	if len(recvMsg.Payload) >= ValidPayloadLength && len(recvMsg.Block) != 0 {
+		if err := rlp.DecodeBytes(recvMsg.Block, preparedBlock); err != nil {
+			return nil, errors.New("[VerifyNewViewMsg] Unparseable prepared block data")
+		}
+
+		blockHash := recvMsg.Payload[:32]
+		preparedBlockHash := preparedBlock.Hash()
+		if !bytes.Equal(preparedBlockHash[:], blockHash) {
+			return nil, errors.New("[VerifyNewViewMsg] Prepared block hash doesn't match msg block hash")
+		}
+		if err := vc.blockVerifier(preparedBlock); err != nil {
+			return nil, errors.New("[VerifyNewViewMsg] Prepared block verification failed")
+		}
+		return preparedBlock, nil
+	}
+	return nil, nil
+}
+
+// VerifyViewChangeMsg returns the type of the view change message after verification
+func (vc *viewChange) VerifyViewChangeMsg(recvMsg *FBFTMessage, members multibls.PublicKeys) (ViewChangeMsgType, *types.Block, error) {
+	vc.vcLock.Lock()
+	defer vc.vcLock.Unlock()
+
+	preparedBlock := &types.Block{}
+	senderKey := recvMsg.SenderPubkey
+	senderKeyStr := senderKey.Bytes.Hex()
+
+	// check and add viewID (m3 type) message signature
+	if _, ok := vc.viewIDSigs[recvMsg.ViewID][senderKeyStr]; ok {
+		return ERR, nil, errors.New("[VerifyViewChangeMsg] received M3 (ViewID) message already")
+	}
+
+	if len(recvMsg.Payload) >= ValidPayloadLength && len(recvMsg.Block) != 0 {
+		if err := rlp.DecodeBytes(recvMsg.Block, preparedBlock); err != nil {
+			return ERR, nil, err
+		}
+		if err := vc.blockVerifier(preparedBlock); err != nil {
+			return ERR, nil, err
+		}
+		_, ok := vc.bhpSigs[recvMsg.ViewID][senderKeyStr]
+		if ok {
+			return M1, nil, errors.New("[VerifyViewChangeMsg] received M1 (prepared) message already")
+		}
+		if !recvMsg.ViewchangeSig.VerifyHash(senderKey.Object, recvMsg.Payload) {
+			return ERR, nil, errors.New("[VerifyViewChangeMsg] failed to verify signature for m1 type message")
+		}
+		blockHash := recvMsg.Payload[:32]
+
+		aggSig, mask, err := chain.ReadSignatureBitmapByPublicKeys(recvMsg.Payload[32:], members)
+		if err != nil {
+			return ERR, nil, err
+		}
+		if !aggSig.VerifyHash(mask.AggregatePublic, blockHash[:]) {
+			return ERR, nil, errors.New("[VerifyViewChangeMsg] failed to verify multi signature for m1 prepared payload")
+		}
+
+		vc.getLogger().Info().Uint64("viewID", recvMsg.ViewID).
+			Str("validatorPubKey", senderKeyStr).
+			Msg("[VerifyViewChangeMsg] Add M1 (prepared) type message")
+		vc.bhpSigs[recvMsg.ViewID][senderKeyStr] = recvMsg.ViewchangeSig
+		vc.bhpBitmap[recvMsg.ViewID].SetKey(senderKey.Bytes, true) // Set the bitmap indicating that this validator signed.
+
+		vc.getLogger().Info().Uint64("viewID", recvMsg.ViewID).
+			Str("validatorPubKey", senderKeyStr).
+			Msg("[VerifyViewChangeMsg] Add M3 (ViewID) type message")
+
+		vc.viewIDSigs[recvMsg.ViewID][senderKeyStr] = recvMsg.ViewidSig
```

Comment:
Missing the viewID sig checking.

Reply:
Also better to keep the old logic flow where 
if m1 {
 do m1 thing
} else {
 do m2 thing
}

// at last
do m3 (viewID) thing

Reply:
m3 is already processed at end end, so the new code processed m1 and return, if not m1, then do m2 and return.  The same logic. ViewID sign checking is moved to consensus/check.go, onViewChangeSanityCheck function.